### PR TITLE
integrate WDB.1 and WDB.2 with generic database server

### PR DESCRIPTION
**Pull request by jni2000**
Please review the integration of WDB.1 and WDB.2 with the generic database server.

### CODE REVIEW

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 1]
```diff
@@ -92,13 +92,20 @@ def generate_WDB_1_default_config(self, filename):
 
     conditionals = [conditionals_b]
 
-    # Create the actual config file
+    pal_database_config = {
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
+        'filePath': os.path.join('testcases', 'testdata', 'wdb_1', 'pal_db_record.json')
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: changed as suggested.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 2]
```diff
@@ -125,21 +132,34 @@ def test_WINNF_FT_S_WDB_1(self, config_filename):
...
-    # Trigger PPA creation
+    # Trigger PPA creation.
     for ppa_creation_request in ppa_creation_requests:
       self.assertPpaCreationFailure(ppa_creation_request)
 
-    # TODO
-    # Step 3: Create PAL database with a PAL record containg the PAL ID 'P'.
+    # Step 3: Create PAL database with a PAL record containing the PAL ID 'P'.
+    pal_database = DatabaseServer('PAL Database',config['palDatabaseConfig']['hostName'],
+                                  config['palDatabaseConfig']['port'], authorization=True)
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: removed

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 3]
```diff
@@ -174,11 +194,17 @@ def generate_WDB_2_default_config(self, filename):
     convertRequestToRequestWithCpiSignature(cpi_private_key_d, cpi_id_d,
                                             cpi_name_d, device_d)
 
-    # Create the actual config.
+    cpi_database_config = {
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
+        'filePath': os.path.join('testcases', 'testdata', 'wdb_2', 'CPI_Database-Public.csv')
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: changed as suggested

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 4]
```diff
@@ -205,24 +231,35 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
...
     for registration_response in registration_responses:
       self.assertEqual(registration_response['response']['responseCode'], 103)
 
     del registration_responses
 
-    # TODO
     # Step 2: Create a CPI database which includes CPI user with credentials
     # matching those used to create the request in Step 1.
+    cpi_database = DatabaseServer('CPI Database',config['cpiDatabaseConfig']['hostName'],
+                                  config['cpiDatabaseConfig']['port'],authorization=True)
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: removed

File: src/harness/testcases/testdata/wdb_1/pal_db_record.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
+{
+  "palId": "pal\/10-2017\/123456789\/A",
+  "licensee": "CBSDOperatorNo.1",
+  "userId": "cbsdOp1",
+  "registrationInformation": {
+    "frn": "123456-789",
+    "registrationDate": "2017-06-01"
+  },
+  "license": {
+    "licenseId": "fccId123456789 - 1",
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/testdata/wdb_1/pal_db_record.json]: removed.

File: src/harness/testcases/testdata/wdb_1/pal_db_record.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    "licenseDate": "2017-07-01T00:00:00Z",
+    "licenseExpiration": "2020-07-01T00:00:00Z",
+    "licenseAreaIdentifier": "CensusTract123456789,Long Name",
+    "licenseAreaExtent": "zone\/census_tract\/census\/2010\/123456789",
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/testdata/wdb_1/pal_db_record.json]: removed

File: src/harness/testcases/testdata/wdb_1/pal_db_record.json [hunk 7]
```diff
@@ -0,0 +1,28 @@
+{
+  "palId": "pal\/10-2017\/123456789\/A",
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/testdata/wdb_1/pal_db_record.json]: removed

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 8]
```diff
@@ -205,24 +231,35 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
...
+
+    # Start CPI database server.
+    cpi_database.start()
+
+    # Set file path.
+    cpi_database.setFileToServe(config['cpiDatabaseConfig']['fileUrl'],
+                                config['cpiDatabaseConfig']['filePath'])
+
+    # Inject the CPI database URL into the SAS UUT.
+    self._sas_admin.InjectDatabaseUrl(config['cpiDatabaseConfig']['fileUrl'])
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: fixed as suggested

File: src/harness/testcases/testdata/wdb_1/pal_db_record.json [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  "licensee": "CBSDOperatorNo.1",
+  "userId": "cbsdOp1",
+  "registrationInformation": {
+    "frn": "123456-789",
+    "registrationDate": "2017-06-01"
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/testdata/wdb_1/pal_db_record.json]: Can you please send me a copy of the PAL record format you referred to? The latest we got from Kate is as below, which is what we used.

{ 
“palId” : “pal/10-2017/123456789/A”, 
“licensee” : “CBSD Operator No.1”, 
“userId” : “cbsdOp1”, 
“registrationInformation” : { 
“frn” : “123456-789”, 
“registrationDate”: “2017-06-01”, 
}, 
“license”: { 
“licenseId”: “fccId123456789-1”, 
“licenseDate”: “2017-07-01T00:00:00Z”, 
“licenseExpiration”: “2020-07-01T00:00:00Z”, 
“licenseAreaIdentifier”: “Census Tract 123456789, Long Name”, 
“licenseAreaExtent”: “zone/census_tract/census/2010/123456789”, 
“licenseFrequencyChannelId”: “1” 
}, 
“licenseStatus”: “VALID”, 
“channelAssignment”: { 
“primaryAssignment”: { 
“lowFrequency”: 3550000000, 
“highFrequency”: 3560000000 
}, 
“secondaryAssignment”: { 
“lowFrequency”: 3580000000, 
“highFrequency”: 3590000000 
} 
} 
}

File: src/harness/testcases/testdata/wdb_2/CPI_Database-Public.csv [hunk 10]
```diff
@@ -0,0 +1,3 @@
+cpiId,status,publicKeyIdentifier 
+ABCdef123,ACTIVE,
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/testdata/wdb_2/CPI_Database-Public.csv]: removed the invalid line in the original file.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 11]
```diff
@@ -92,13 +92,20 @@ def generate_WDB_1_default_config(self, filename):
 
     conditionals = [conditionals_b]
 
-    # Create the actual config file
+    pal_database_config = {
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: fixed using the format of FDB.3

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 12]
```diff
@@ -174,11 +194,17 @@ def generate_WDB_2_default_config(self, filename):
     convertRequestToRequestWithCpiSignature(cpi_private_key_d, cpi_id_d,
                                             cpi_name_d, device_d)
 
-    # Create the actual config.
+    cpi_database_config = {
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: fixed using the format of FDB.3

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: I do not recall that we will have multiple CPI database files. Can you please point me to the document or reference ? Thanks.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 13]
```diff
@@ -205,24 +231,35 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
...
     # matching those used to create the request in Step 1.
+    cpi_database = DatabaseServer('CPI Database',config['cpiDatabaseConfig']['hostName'],
+                                  config['cpiDatabaseConfig']['port'],authorization=True)
+
+    # Start CPI database server.
+    cpi_database.start()
+
+    # Set file path.
+    cpi_database.setFileToServe(config['cpiDatabaseConfig']['fileUrl'],
+                                config['cpiDatabaseConfig']['filePath'])
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: fixed as suggested.

File: src/harness/testcases/testdata/wdb_2/CPI_Database-Public.csv [hunk 14]
```diff
@@ -0,0 +1,3 @@
+cpiId,status,publicKeyIdentifier 
+ABCdef123,ACTIVE,
+123456,ACTIVE,https://winnf.memberclicks.net/mcdatafiles/receiptattach/winnf/12764702/8960734/Public_Key_Dummy_File.txt
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/testdata/wdb_2/CPI_Database-Public.csv]: The CPI database file is provided by the organization with authority which will be WinnForum in this case. I think we shall keep what we have from the WinnForum. Please advise if you think otherwise

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 15]
```diff
@@ -174,11 +194,17 @@ def generate_WDB_2_default_config(self, filename):
     convertRequestToRequestWithCpiSignature(cpi_private_key_d, cpi_id_d,
                                             cpi_name_d, device_d)
 
-    # Create the actual config.
+    cpi_database_config = {
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
+        'filePath': os.path.join('testcases', 'testdata', 'wdb_2', 'CPI_Database-Public.csv')
+    }
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Is this related to the multiple CPI file comments ? If so, we will address the needs when that one is clarified.

**Comment by jni2000** [comment]
Taran

We have addressed most of your review comments except the followings
1. the private key file issue -- we are working on it
2. the PAL record format -- please can you share with me the format you have ? I believe we used the latest from Kate.
3. the three comments related to supporting multiple CPI files, I do not recall we have such a requirements and I am calcifying this with Kate right now. 

Please take a look of others and let's move on. Thanks.

**Comment by kate-harrison** [comment]
2. Please use the format here: https://workspace.winnforum.org/higherlogic/ws/public/download/4488/WINNF-TS-0245-V1.0.0%20PAL%20Database%20.pdf

3. Yes, this is a requirement.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kate-harrison) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Confirmed, you should inject the URL to the index here.

File: src/harness/testcases/testdata/wdb_1/pal_db_record.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (kate-harrison) [reply src/harness/testcases/testdata/wdb_1/pal_db_record.json]: https://workspace.winnforum.org/higherlogic/ws/public/download/4488/WINNF-TS-0245-V1.0.0%20PAL%20Database%20.pdf

**Comment by jni2000** [comment]
2. Just double checked, the pal record file we used is same as the one in https://workspace.winnforum.org/higherlogic/ws/public/download/4488/WINNF-TS-0245-V1.0.0%20PAL%20Database%20.pdf.
3. Is there a specific reason that we need to treat this different than the PAL database file we discussed recently ? I would propose we keep them consistent.

File: src/harness/testcases/testdata/wdb_2/WDB_2_CPI_Private_Key.txt [hunk 16]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
[the remark this answers is not in the document]
Comment (kate-harrison) [reply src/harness/testcases/testdata/wdb_2/WDB_2_CPI_Private_Key.txt]: Confirmed, only the public key should be distributed. We don't need to even store the private key in the repo.

**Comment by kate-harrison** [comment]
2. It appears as though the example in that file is out of date. Please refer to the end of page 5: "Date of licensee FCC registration. It is expressed using the format, YYYYMM-DDThh:mm:ssZ, as defined by [N.6]."

3. Yes, we need to serve the public key files in addition to the index, whereas the PAL DB has only one file it needs to serve.

**Comment by jni2000** [comment]
Thanks, Kate. Fixed the PAL record already. For helping us speed up, please can you see if my understanding about the multiple CPI files is right:

1. the CPI file will be of the format as defined in the file from the WinnF link
2, we will have multiple of them and the test codes just need to put them in a indexed order for SAS UUT to retrieve.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TaranFurnell) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: The test CPI database has to host both the database index file (the one that the SAS is informed about), as well as the individual private keys for each CPI (which it gets the url for from the database).

After a bit more discussion with Kate we have the following suggested solution:

As part of the CPI database config have a field `cpis` set similar to the following
```python
'cpis': [
    {'id': 'abcd', 'status': 'ACTIVE', 'public_key_file': 'some/file/path'},
    {'id': 'defg', 'status': 'ACTIVE', 'public_key_file': 'some/other/file/path'}
]
```

and the `indexUrl` as already asked for.

Then in the test case itself use this information to create the index file, each line looks something like:
```python
cpi['id'] + ',' + cpi['status'] + ',/' + cpi['public_key_file']
```

and set the contents of the database server to be:
```python
files = {('/'+file['public_key_file']): file['public_key_file'] for file in config[cpis]}
files[config['indexUrl']] = # whatever file path the index file was saved at.
cpi_database.setFilesToServe(files)
```

**Comment by TaranFurnell** [comment]
See the comment I just wrote, the gist is a there is the index file which the SAS will be informed about. Then each CPI public key that is referred to by that index file must also be provided by this server.

**Comment by jni2000** [comment]
what will be the index file format ? Is there a document on this ? Will it be like a table of rows of the format like "index, cpiFileName" ?

**Comment by jni2000** [comment]
Taran and Kate, we have addressed all the review comments and made necessary changes, please review.

Some special notes:
1. In the CPI test case, we need both the CPI public key and private key, but for different purposes
2. the public key is part of the database data to be retrieved by SAS UUT for decrypting the CPI signed data and thus put into the database data folder "/src/harness/testcases/testdata/cpi_db"
3. the private key is used for the test harness codes to encrypting the CPI data when mimicking the CBSD with CPI signed data, this is test data and thus put into "/src/harness/testcases/testdata/"

This is what we believe the right arrangement, please advise if you think otherwise.

Thanks.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 17]
```diff
@@ -92,13 +92,20 @@ def generate_WDB_1_default_config(self, filename):
 
     conditionals = [conditionals_b]
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: I think it makes the most sense to save the PAL records above to a file, then use that file in the PAL DB config. Otherwise it's not clear that the CBSDs and the PAL records will be self-consistent (same user ID, right license area, etc.). I suggest that you take the PAL records created above, save them to a file during the default config generation (same format as the file you already have), then give that filename as part of the default config.

The same applies to the CPI information, since you will only need the private keys when creating the config (and not during the test case itself).

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 18]
```diff
@@ -205,24 +233,38 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
     # Send registration requests.
     registration_responses = self._sas.Registration(registration_requests)['registrationResponse']
 
-    # Check registeration response,
+    # Check registration response,
     # responseCode should be 103 (INVALID_VALUE).
     for registration_response in registration_responses:
-      self.assertEqual(registration_response['response']['responseCode'], 103)
+      self.assertEqual(registration_response['response']['responseCode'], 130)
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Please revert.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 19]
```diff
@@ -205,24 +233,38 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
...
     # Step 2: Create a CPI database which includes CPI user with credentials
     # matching those used to create the request in Step 1.
+    cpi_database = DatabaseServer('CPI Database',config['cpiDatabaseConfig']['hostName'],
+                                  config['cpiDatabaseConfig']['port'])
+
+    # Start CPI database server.
+    cpi_database.start()
+
+    # Prepare the file list.
+    files = {('/' + file['public_key_file']): file['public_key_file'] for file in config['cpiDatabaseConfig']['cpis']}
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Please don't use the var name `file` since it is a Python keyword. Probably `cpi` makes more sense here anyway.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kate-harrison) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: I think you missed part of this: the index file can be written in the test case itself, and does not need to be committed to GitHub.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 17 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Updated as per suggestion.
Thanks for the suggestion. This is a good way of making CBSDs and PAL records consistent and use the same in the database.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 18 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Reverted.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 19 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Fixed as per suggestion.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Fixed as per suggestion.
To generate CSV file, I have used the first instance of cpis information like below 
config['cpiDatabaseConfig']['cpis'][0]['publicKeyIdentifierFile'] 
Please let me know this is correct way of doing it?

**Comment by janetliuboston** [comment]
Hi, This is ready for review again. Thanks!

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 20]
```diff
@@ -92,13 +93,26 @@ def generate_WDB_1_default_config(self, filename):
 
     conditionals = [conditionals_b]
 
-    # Create the actual config file
+    pal_db_relative_file_path = os.path.join('testcases', 'testdata', 'pal_db', 'pal_db_record.json')
+
+    # Create the pal_db record file consistent with CBSDs and PAL records.
+    with open(os.path.join(pal_db_relative_file_path), 'w+') as file_handle:
+      file_handle.write(json.dumps(pal_records_a,pal_records_b,indent=2))
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: As written, this will write a list of lists to the JSON file. Instead, we want just one list of PAL DB records.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 21]
```diff
@@ -92,13 +93,26 @@ def generate_WDB_1_default_config(self, filename):
 
     conditionals = [conditionals_b]
 
-    # Create the actual config file
+    pal_db_relative_file_path = os.path.join('testcases', 'testdata', 'pal_db', 'pal_db_record.json')
+
+    # Create the pal_db record file consistent with CBSDs and PAL records.
+    with open(os.path.join(pal_db_relative_file_path), 'w+') as file_handle:
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Why do you need another call to os.path.join() here?

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 22]
```diff
@@ -160,11 +187,11 @@ def generate_WDB_2_default_config(self, filename):
     cpi_name_d = 'd_name'
 
     # Read private keys for the CPI users
-    with open(os.path.join('testcases', 'testdata', 'wdb_2', 'WDB_2_CPI_Private_Key.txt'),
+    with open(os.path.join('testcases', 'testdata','WDB_2_CPI_Private_Key.txt'),
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: (1) You should be generating the public+private key pair in this file (see the REG test cases for an example, it's a single function call), not reading them from disk.
(2) You need write out only the public key, NOT the private key.
(3) Line 194 uses the same file. The CPIs need to have distinct keys.
(4) As you did above, please save the filename in a local var and then use that directly in the config.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 23]
```diff
@@ -174,11 +201,25 @@ def generate_WDB_2_default_config(self, filename):
...
-    # Create the actual config.
+    cpi_database_config = {
+        'hostName': 'localhost',
+        'port': 8003,
+        'cpis': [{
+            'cpiId': cpi_id_b,
+            'status': 'ACTIVE',
+            'publicKeyIdentifierFile': os.path.join('testcases', 'testdata', 'cpi_db', 'CPI_Database-Public.csv')},
+            {
+                'cpiId': cpi_id_d,
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Please use consistent indentation for these two CPI dicts so that it's clear they're in the same list.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 24]
```diff
@@ -205,24 +246,48 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
...
+                                  config['cpiDatabaseConfig']['port'])
+
+    # Start CPI database server.
+    cpi_database.start()
+
+    # Generate the CSV file and write the content of it.
+    config_keys = config['cpiDatabaseConfig']['cpis'][0].keys()
+
+    with open(config['cpiDatabaseConfig']['cpis'][0]['publicKeyIdentifierFile'], 'w+') as output_file:
+        dict_writer = csv.DictWriter(output_file, config_keys)
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: This will not write them in a deterministic order, since line 265 is not deterministic. (Moreover, the dict keys are not the same as the column names.) Please hard-code the column names.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 25]
```diff
@@ -205,24 +246,48 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
...
+        dict_writer.writeheader()
+        dict_writer.writerows(config['cpiDatabaseConfig']['cpis'])
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: This will not produce full URLs for the PKI file, which is needed.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 26]
```diff
@@ -205,24 +246,48 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
...
+
+    # Prepare the file list.
+    files = {('/' + cpi['publicKeyIdentifierFile']):
+                 cpi['publicKeyIdentifierFile'] for cpi in config['cpiDatabaseConfig']['cpis']}
+    files[config['cpiDatabaseConfig']['indexUrl']] = \
+        os.path.join('testcases', 'testdata', 'cpi_db', 'indexUrl.csv')
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Let's call it just "index.csv".

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py [hunk 27]
```diff
@@ -205,24 +246,48 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
...
+    cpi_database = DatabaseServer('CPI Database',config['cpiDatabaseConfig']['hostName'],
+                                  config['cpiDatabaseConfig']['port'])
+
+    # Start CPI database server.
+    cpi_database.start()
+
+    # Generate the CSV file and write the content of it.
+    config_keys = config['cpiDatabaseConfig']['cpis'][0].keys()
+
+    with open(config['cpiDatabaseConfig']['cpis'][0]['publicKeyIdentifierFile'], 'w+') as output_file:
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: This is the wrong file. You should be writing index.csv (which is then used on line 275), not overwriting the PKI files.

Please store the path in a local var and use that var below, to reduce the chance for error.

Please call the file index.csv.

File: src/harness/testcases/testdata/WDB_2_CPI_Private_Key.txt [hunk 28]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (kate-harrison) [inline src/harness/testcases/testdata/WDB_2_CPI_Private_Key.txt]: Not needed.

File: src/harness/testcases/testdata/cpi_db/CPI_Database-Public.csv [hunk 29]
```diff
@@ -0,0 +1,2 @@
+cpiId,status,publicKeyIdentifier
```
Comment (kate-harrison) [inline src/harness/testcases/testdata/cpi_db/CPI_Database-Public.csv]: Why is this here?

File: src/harness/testcases/testdata/cpi_db/WDB_2_CPI_Private_Key.txt [hunk 30]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (kate-harrison) [inline src/harness/testcases/testdata/cpi_db/WDB_2_CPI_Private_Key.txt]: Not needed.

File: src/harness/testcases/testdata/pal_db/pal_db_record.json [hunk 31]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (kate-harrison) [inline src/harness/testcases/testdata/pal_db/pal_db_record.json]: This should be a list of PAL DB records (i.e. there should be `[]` on the outermost level). Please ensure that the file generated by the default config follows that format, but please do not commit that file.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 20 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: We have doubled checked it. This statement 
json.dumps(pal_records_a,pal_records_b,indent=2)
write a list of PAL records.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 21 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Not required, hence removed.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 23 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Fixed as per suggestion.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 22 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Fixed as per suggestion.

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 24 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Fixed as per suggestion

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 25 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Fixed as per suggestion

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 26 above)
Comment (janetliuboston) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Fixed as per suggestion

File: src/harness/testcases/testdata/WDB_2_CPI_Private_Key.txt (hunk 28 above)
Comment (janetliuboston) [reply src/harness/testcases/testdata/WDB_2_CPI_Private_Key.txt]: WDB_2_CPI_Private_Key.txt is removed

File: src/harness/testcases/testdata/cpi_db/CPI_Database-Public.csv (hunk 29 above)
Comment (janetliuboston) [reply src/harness/testcases/testdata/cpi_db/CPI_Database-Public.csv]: WDB_2_CPI_Private_Key.txt is removed

File: src/harness/testcases/testdata/pal_db/pal_db_record.json (hunk 31 above)
Comment (janetliuboston) [reply src/harness/testcases/testdata/pal_db/pal_db_record.json]: We have doubled checked it. This statement 
json.dumps(pal_records_a,pal_records_b,indent=2)
write a list of PAL records.

**Comment by janetliuboston** [comment]
Updated with the latest. Please review. Thanks!

File: src/harness/testcases/WINNF_FT_S_WDB_testcase.py (hunk 20 above)
Comment (kate-harrison) [reply src/harness/testcases/WINNF_FT_S_WDB_testcase.py]: Are you sure that `pal_records_b` is written out? I did an experiment locally and it doesn't seem to be the case.

**Comment by kate-harrison** [comment]
I think we can close this now, yes?

**Comment by janetliuboston** [comment]
Yes, this is superseded by https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/599
